rfcs/0001-meta-schedule-autotensorir.md

```diff
@@ -0,0 +1,444 @@
+    j_0, j_1, j_2, j_3 = sch.split(loop=j, factors=j_tiles)
+    k_0, k_1 = sch.split(loop=k, factors=k_tiles)
+    # Organize the loops into "SSRSRS" 6-level tiles
+    sch.reorder(
+        i_0, j_0, # S
+        i_1, j_1, # S
+        k_0,      # R
+        i_2, j_2, # S
+        k_1,      # R
+        i_3, j_3, # S
+    )
+    return sch
+```
+
+**Customize composite schedule in python**. We provide two ways to define a composite schedule in
+python:
+
+Method 1. Derive from `PyCompositeSchedule`, and implement two methods `initialize` and `apply`:
+
+```python
+class MultiLevelTiling(PyCompositeSchedule):
+    def initialize(...):
+        # initialize the class, usually this method is empty
+        ...
+
+    def apply(sch: Schedule, block: BlockRV) -> Union[Schedule, List[Schedule]]:
+        # write any python code, including:
+        # - analyze `block`
+        # - invoke schedule primitives in `sch`
+        # - do debug printing
+        ...
+```
+
+Method 2. A decorator as the syntactic sugar if the `initialize` method is empty, which converts the
+function to the `apply` method.
+
+```python
+@tir.as_composite_schedule(name="multi-level-tiling")
+def multi_level_tiling(sch: Schedule, block: BlockRV) -> Union[Schedule, List[Schedule]]:
+    # write any python code, including:
+    # - analyze `block`
+    # - invoke schedule primitives in `sch`
+    # - do debug printing
+    ...
+```
+
+**Customize exploration strategies in python**. Developers can implement any search algorithm in
+python as well by deriving from `PySearchPolicy`, and the syntax is identical to customizing with
+`PyCompositeSchedule`.
+
+**Other customizable components**. This list includes:
+
+* Cost model
+* Database
+* Measure callbacks
+* Feature extractor
+* Program builder & runner
+* Analysis methods
+* ...
+
+In a short summary, almost every component of the system is decoupled with each other and extensions
+could be easily plugged in.
+
+## 5. Drawbacks
+
+We are not aware of any drawbacks of the proposed system.
+
+## 6. Rationale and alternatives
+
+The system is designed with the principle of minimalism: different from alternative solutions, we do
+not require any change in existing codebase, or extra APIs to learn. It could potentially lower the
+bar of using automation systems. 
+
+Unifying manual scheduling, AutoTVM's semi automatic templates and AutoScheduler's (Ansor's) fully
+automatic sketch generation provides flexible way to balance injection new domain knowledge and
+automation.
+
+Flexibility in customization allows quick try-out on new tasks, new strategies and new hardware
+targets without deep knowledge of the system.
+
+## 7. Prior art
+
+**Tensor Expression (TE)** in TVM is a DSL that decouples compute and schedule, which provides
+convenient ways to handcraft optimized kernels for different hardware targets.
+
+**TensorIR** is the latest generation of TVM’s low-level IR. Its capability of eagerly applying
```

Comment:
is this different from TIR? I don't think you should make a v2 of TIR called TensorIR, because I already assumed TIR was short for TensorIR in previous materials.

Reply:
We should assume TIR is short for TensorIR in all the materials. Perhaps I should remove "the latest generation" here